strongbox-storage/strongbox-storage-api/src/main/java/org/carlspring/strongbox/services/impl/ArtifactEntryServiceImpl.java

```diff
@@ -81,52 +103,77 @@ private boolean artifactEntryIsSavedForTheFirstTime(ArtifactEntry artifactEntry)
         return save(entity, false);
     }
 
-    private <S extends ArtifactEntry> void updateLastVersionTag(S entity)
+    private <S extends ArtifactEntry> void updateLastVersionTag(S entity,
+                                                                ArtifactGroup artifactGroup)
     {
         ArtifactCoordinates coordinates = entity.getArtifactCoordinates();
         Assert.notNull(coordinates, "coordinates should not be null");
-        
+
         ArtifactTag lastVersionTag = artifactTagService.findOneOrCreate(ArtifactTagEntry.LAST_VERSION);
-        Set<ArtifactTag> tagSet = new HashSet<>();
-        tagSet.add(lastVersionTag);
-        
-        Map<String, String> coordinatesMap = coordinates.dropVersion();
-        
-        List<ArtifactEntry> lastVersionArtifactList = findArtifactList(entity.getStorageId(),
-                                                                       entity.getRepositoryId(),
-                                                                       coordinatesMap, tagSet,
-                                                                       0,
-                                                                       -1,
-                                                                       "uuid", true).stream()
-                                                                                    .map(e -> (ArtifactEntry) getDelegate().detachAll(e,
-                                                                                                                                      true))
-                                                                                    .collect(Collectors.toList());
-        ArtifactEntry lastVersionEntry = lastVersionArtifactList.stream().findFirst().orElse(entity);
-        Optional<ArtifactCoordinates> lastVersionCoordinates = Optional.ofNullable(lastVersionEntry.getArtifactCoordinates());
-        if (lastVersionEntry.equals(entity))
+
+        Set<ArtifactEntry> lastVersionEntries = findLastVersionArtifactEntries(artifactGroup,
+                                                                               lastVersionTag,
+                                                                               entity);
+        lastVersionEntries.stream()
+                          .forEach(lastVersionEntry -> {
+                              setLastVersionTag(lastVersionEntry, entity, lastVersionTag);
+                          });
+    }
+
+    private <S extends ArtifactEntry> void setLastVersionTag(S lastVersionEntry,
```

Comment:
maybe we should call this method `checkAndUpdateLastVersionTagIfNeeded`? :)
I mean that probably there should be clearer name for this method, @fuss86  what do you think?

Reply:
I like most expressed method names so I agree :)